NSSCodeForNashville/NashvilleHistory
# Pretty big refactor, also create, edit and delete Tours/Routes

## Status
**READY**

## Documentation

[X] I have fully documented features in this PR's code. Please put an `X` in the box to confirm.

[ ] This PR does **NOT** require documentation.

## Description
The original purpose of this branch was to create, edit and delete Custom Routes (often called `Tours` throughout the code). In completing this task, I refactored several areas of the app for efficiency and reliability.

Trying to maintain a dry data structure, I wanted to store places on each Tour by reference, but found that Nashville API doesn't provide a guid for each marker. To alleviate this, I calculated a string uid for each place on retrieval that is a combination of the first word in the title, followed by full latitude and longitude without any periods or minus signs.

Since we don't have a way to retrieve markers from the Nashville API by uid (retrieval by latitude/longitude would be unreliable for reference). I opted to store a local copy of all markers in client memory in the top-level AppCtrl, so marker uid would be available to all sections of the app. This results in all markers being pulled down on page load, and then each controller simply manipulates the dataset available from its parent controller.

Since we're now pulling down about 450 markers on app load (instead of markers within a certain radius), I refactored the Distance calls to reduce our usage of the Google API. Now, MarkersCtrl gets a distance for all markers manually, sorts them, then confirms that distance through Google Distance Matrix for the closest ten markers only.

Highlights:
 * Logged-In Users can create custom Tours and add places
    * They can also delete tours, delete and reorder places on those tours
 * All Markers are being retrieved one time on app load
    * Each Place is assigned a uid based on title, latitude and longitude
    * AppCtrl.AllPlaces is an array with all places that never changes
    * AppCtrl.MarkerCards is an array that holds only places being displayed to the user
 * MarkersCtrl calculates distance for all markers manually, then confirms the closest 10 with Google API and resorts

## Resolves Issue Number
Resolves #28 

## Deploy Notes
Watch out for CORS

```sh
git pull --prune
git checkout kraatz-tour
ionic serve --lab
```

## Steps to Test or Reproduce

1. Exercise the Add to Route and My Tours features
1. Check for strange behavior around the rest of the app

## Impacted Files in Application
See diff and in-line comments
